### PR TITLE
seal: Refactor ext_gas_price

### PR DESCRIPTION
**Pull request by athei**
Subset of: https://github.com/paritytech/substrate/issues/6355

# Overview

This PR changes the signature of `ext_gas_price` from
```rust
fn ext_gas_price() -> BalanceOf<T>
```

to
```rust
fn ext_gas_price(gas: Gas) -> BalanceOf<T>
```

In addition to that we remove the dependency of seal on `pallet_transaction_payment` by adding those two elements to the `Trait`:
```rust
/// The currency in which fees are paid and contract balances are held.
type Currency: Currency<Self::AccountId>;

/// Used to supply contracts with the current gas price. This is **not** used to
/// calculate the actual fee and is only for informational purposes.
type WeightPrice: Convert<Weight, BalanceOf<Self>>
```

# Motivation
There is no longer something like gas. We merely keep the term around for historical reasons. Costs are measured in weight. Currently, we provide gas price as the costs for one weight. Calculating the costs for a single weight is subject to rounding errors because the costs for one weight can be below one. In addition to that a fixed point adjustment is applied to the fee. Both of that makes calculating the costs for one weight very brittle. Contracts should therefore pass in higher amounts of gas to get a useful result.

### CODE REVIEW

File: frame/contracts/src/exec.rs [hunk 1]
```diff
@@ -205,7 +206,7 @@ pub trait Ext {
 	fn get_runtime_storage(&self, key: &[u8]) -> Option<Vec<u8>>;
 
 	/// Returns the price of one weight unit.
-	fn get_weight_price(&self) -> BalanceOf<Self::T>;
+	fn get_weight_price(&self, weight: Weight) -> BalanceOf<Self::T>;
```
Comment (pepyakin) [inline frame/contracts/src/exec.rs:209]: Comment update needed.

File: frame/contracts/src/lib.rs [hunk 2]
```diff
@@ -373,6 +378,10 @@ pub trait Trait: frame_system::Trait + pallet_transaction_payment::Trait {
 
 	/// The maximum size of a storage value in bytes.
 	type MaxValueSize: Get<u32>;
+
+	/// Used to supply contracts with the current gas price. This is **not** used to
```
Comment (pepyakin) [inline frame/contracts/src/lib.rs:382]: Prob. we should put it as "used to answer contracts' queries regarding the current gas/weight price"

File: frame/contracts/src/exec.rs (hunk 1 above)
Comment (athei) [reply frame/contracts/src/exec.rs:209]: Done.

File: frame/contracts/src/lib.rs (hunk 2 above)
Comment (athei) [reply frame/contracts/src/lib.rs]: That is better. Done.

File: frame/contracts/src/wasm/mod.rs [hunk 3]
```diff
@@ -375,8 +376,8 @@ mod tests {
 				)
 			)
 		}
-		fn get_weight_price(&self) -> BalanceOf<Self::T> {
-			1312_u32.into()
+		fn get_weight_price(&self, weight: Weight) -> BalanceOf<Self::T> {
+			(weight * 1312).into()
```
Comment (niklasad1) [inline frame/contracts/src/wasm/mod.rs:380]: why not `saturating_mul`?

Comment (athei) [reply frame/contracts/src/wasm/mod.rs]: I didn't bother because it's test code. It does not hurt, though. I switched to `saturating_mul`.

Comment (niklasad1) [reply frame/contracts/src/wasm/mod.rs]: Ok, I didn't notice that :)

**Comment by pepyakin** [comment]
bot merge

**Pull request merged**